scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,273 @@
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow;
+use chrono::Duration;
+use core::time;
+use futures::stream::{FuturesUnordered, StreamExt};
+use scylla::frame::response::result::Row;
+use scylla::Session;
+use tokio::sync::mpsc;
+use tokio::time::sleep;
+
+use scylla_cdc::cdc_types::GenerationTimestamp;
+use scylla_cdc::reader::StreamReader;
+use scylla_cdc::stream_generations::GenerationFetcher;
+
+pub struct CDCLogPrinter {
+    session: Arc<Session>,
+    start_timestamp: Duration,
+    end_timestamp: Duration,
+    sleep_interval: time::Duration,
+    window_size: Duration,
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+}
+
+impl CDCLogPrinter {
+    pub fn new(
+        session: &Arc<Session>,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+    ) -> CDCLogPrinter {
+        CDCLogPrinter {
+            session: Arc::clone(session),
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            readers: vec![],
+        }
+    }
+
+    pub async fn run(&mut self, keyspace: String, table_name: String) -> anyhow::Result<()> {
+        let fetcher = GenerationFetcher::new(&self.session);
+        let mut generation = self.get_first_generation(&fetcher).await?;
+
+        self.readers = vec![];
+        for stream_id in fetcher
+            .fetch_stream_ids(&generation)
+            .await?
+            .iter()
+            .flatten()
+        {
+            let reader = StreamReader::new(
+                &self.session,
+                vec![stream_id.clone()],
+                self.start_timestamp,
+                self.window_size,
+                self.safety_interval,
+                self.sleep_interval,
+            );
+            self.readers.push(Arc::new(reader));
+        }
+
+        fn print_row(row: Row) {
+            println!("{}", row.columns.len());
+        }
+
+        let mut futures: FuturesUnordered<_> = self
+            .readers
+            .iter()
+            .map(|reader| {
+                let reader = Arc::clone(reader);
+                let keyspace = keyspace.clone();
+                let table_name = table_name.clone();
+                tokio::spawn(async move { reader.fetch_cdc(keyspace, table_name, print_row).await })
+            })
+            .collect();
+
+        let (generation_sender, mut generation_receiver) = mpsc::channel(1);
+
+        let sleep_interval = self.sleep_interval;
+        let end_timestamp = self.end_timestamp;
+        tokio::spawn(async move {
```

Comment:
Suggestion: it might also make sense (or not) to move the `get_first_generation` logic to the beginning of the task so that the first generation is pushed to the generations channel, too, and can be handled in a unified way.